x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingAppender.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.deprecation.logging;
+
+import org.apache.logging.log4j.core.Appender;
+import org.apache.logging.log4j.core.Core;
+import org.apache.logging.log4j.core.Filter;
+import org.apache.logging.log4j.core.Layout;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.appender.AbstractAppender;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * This log4j appender writes deprecation log messages to an index. It does not perform the actual
+ * writes, but instead constructs an {@link IndexRequest} for the log message and passes that
+ * to a callback.
+ */
+@Plugin(name = "DeprecationIndexingAppender", category = Core.CATEGORY_NAME, elementType = Appender.ELEMENT_TYPE)
+public class DeprecationIndexingAppender extends AbstractAppender {
+    public static final String DEPRECATION_MESSAGES_DATA_STREAM = "logs-deprecation-elasticsearch";
```

Comment:
this is going to match the default logs-*-* template which defines the following mappings:

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/resources/logs-mappings.json

We should make sure that we populate the predefined mappings for this template. (datastream, ecs, host, etc.)

Reply:
this at the moment depends on the `ECSJsonLayout` where we use https://github.com/elastic/ecs-logging-java to generate ECS compliant JSON. The same JSON is written to our json logs. 
A sample is available here https://github.com/elastic/beats/blob/master/filebeat/module/elasticsearch/deprecation/test/es_deprecation-json.800.log
(note this could utilise ECS even more by renaming `node.name` to `service.node.name`)


Adding ecs version and host would probably be worthy to add to these json logs as well. I am not sure we should populate `datastream` to a JSON written to logs. 

My point is, that this difference would mean we would need a separate layout for `JSON logs` and for `indexed logs` ?
Is it worthy to make other ES logs (server, slow logs) following the `logs-mappings.json` ?

Reply:
> Is it worthy to make other ES logs (server, slow logs) following the logs-mappings.json ?

The goal of the `logs-mapping` that ships by default is to have a basic baseline of mappings.  We can always introduce more specific ones as needed. w.r.t to these deprecation logs we should probably wait for some usage to add customizations beyond the baseline. 

> I am not sure we should populate datastream to a JSON written to logs.

I think it would be OK to include this just index request.

Reply:
@jakelandis @pgomulka so do I need to make any changes here?

Reply:
I am +1 to add the datastream information and the host.ip, but don't think it is a blocker to this PR. (doing a last round of review right now)

Reply:
@jakelandis so we do populate more fields at the moment (node.id, node.name, cluster.uuid, etc) Should we add them to the `logs-mapping` ? 

@pugnascotia I think we need to add datastream and ecs version fields to the log message.